### ISSUE TITLE
Expose token positions for local variables in service protocol

### ISSUE THREAD

**Issue by johnmccutchan** [dart-lang/sdk#25569]

**Comment by johnmccutchan** [comment]
I'd like 'Jump to Source' to work also for local variables and method arguments. In this case I have BoundVariable in hands, but there's no legal way to get info about its position in source. Any chance to add something like BoundVariable.getLocation() to the VM Service protocol? Let me know if you want me to file a request.

**Comment by jwren** [comment]
@johnmccutchan Friendly ping.

Having this would enable us to have significantly better debugging experience in the Dart Plugin for IntelliJ with Inline Debugging: https://www.jetbrains.com/help/idea/2016.1/inline-debugging.html?origin=old_help

**Comment by johnmccutchan** [comment]
Hi @jwren no work on this has been done so far.

**Comment by devoncarew** [comment]
@alexander-doroshko, is the above comment correct in terms of the additional info we'll need from the debugging protocol in order to support https://youtrack.jetbrains.com/issue/WEB-1612?

**Comment by alexander-doroshko** [comment]
Yes, this issue is filed on behalf of me and description (first comment) is a quote from my email.

**Comment by johnmccutchan** [comment]
@alexander-doroshko @jwren @devoncarew This bug needs to be updated with specific examples of the data you'd like over the service protocol. Once this happens I will start work on implementing support for this in the VM.

**Comment by alexander-doroshko** [comment]
@johnmccutchan, the task is to be able to get line and column of the local variables and function parameters. I guess there could be different solutions. 
My suggestion is to add `getLocation()` method to the `BoundVariable` class that would return an instance of `SourceLocation` class pointing to the local var declaration / function parameter declaration.
There's already `Field.getLocation()` method that gives location of the field declaration in the source code (for example if field is declared as `var foo = 1;` this method points to `foo` token. So I thought about the same approach for `BoundVariable`s. If `BoundVariable` represents `this`, its `getLocation()` could be `null`.

**Comment by devoncarew** [comment]
For context, we're using the `Field.getLocation()` information here: https://github.com/JetBrains/intellij-plugins/blob/master/Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/frame/DartVmServiceValue.java#L77.

**Comment by johnmccutchan** [comment]
Again, I need concrete examples of the data you need to implement support for this feature in IntelliJ. I am not interested or asking for API designs.

**Comment by alexander-doroshko** [comment]
Ok let me try to formulate it as an example of data returned by Observatory:
- Isolate1 is suspended
- Execution stack has 9 frames
- Frame 5 contains vars `this`, `foo`, `bar`
- `foo` is a local variable declared in script `file://my/project/bin/main.dart` at line 15, column 10
- `bar` is a function parameter declared in script `file://my/project/bin/main.dart` at line 14, column 21

**Comment by johnmccutchan** [comment]
@alexander-doroshko So you need the line / column that a parameter / variable is declared at.

What about when a local variable goes out of scope?

What precision is needed here? I believe Dart reports a local variable as existing after it's initializing expression, i.e.:

var x = 99;

"x" exists only *after* 99. Is that sufficient? Or do you need it to be the exact position in the source that "x" is declared at?

**Comment by alexander-doroshko** [comment]
> What about when a local variable goes out of scope?

I thought variables that belong to a frame in a suspended stack can't be out of scope, can they? If they can and position info is not known, it can be skipped.

> "x" exists only after 99. Is that sufficient?

I'm afraid not. We need a pointer to a variable name identifier, not to `var` keyword, var initializer or semicolon. I noticed hidden fields of BoundVariable (_tokenPos and _endTokenPos), but it's not exactly what we need.

> Or do you need it to be the exact position in the source that "x" is declared at?

Yes, that's what we need. In this case on 'Jump to source' IDE will move caret exactly to the `x` in `var x = 99;`.

**Comment by johnmccutchan** [comment]
@alexander-doroshko 

Consider:

```dart
func(x, y, z) {
  for (var i = 0; i < 10; i++) {
  }
  { // Explicit scope
    var j = 4;
  }
}
```

`i` is only alive during the for loop and `j` is only alive during the `// explicit scope`.

**Comment by alexander-doroshko** [comment]
Sorry, what is the purpose of this example?
[Frame.getVars()](https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/service.md#frame) doesn't return local vars which are not visible at the suspended point. In the example below we get only `x`,`y` аnd `z`.

![image](https://cloud.githubusercontent.com/assets/4499789/19251496/37e55dde-8f48-11e6-8fdb-b1a4a36fcac7.png)

**Comment by johnmccutchan** [comment]
@alexander-doroshko That is true, so the list of local variables is already filtered.

So what you need is as follows:

For every local variable or function parameter -> the token position of the variable name as described in the source code.

**Comment by alexander-doroshko** [comment]
> For every local variable or function parameter -> the token position of the variable name as described in the source code.

Yup! :thumbsup:

**Comment by johnmccutchan** [comment]
All `BoundVariables` now have:

* `declarationTokenPos` -- position in source where variable name is declared.
*  `scopeStartTokenPos` -- position in source where the variable enters scope.
* `scopeEndTokenPos` -- position in source where the variable exits scope.

**Comment by devoncarew** [comment]
I'm re-generating the Java library now, and will commit it to the intellij repo.

**Comment by alexander-doroshko** [comment]
:+1:

**Comment by alexander-doroshko** [comment]
@johnmccutchan I've added support at IDE side and it works like a charm!
One question: does it make sense to return `declarationTokenPos` for `BoundVariable` that represents `this`?
